### PR TITLE
applications: asset_tracker: AT commands from nRF Cloud

### PR DESCRIPTION
**Pull request by jayteemo**
Implemented receving modem AT commands from the cloud
and returning the modem's response.

Signed-off-by: Justin Morton <justin.morton@nordicsemi.no>

### CODE REVIEW

**Comment by jayteemo** [comment]
good point about the workqueues, i'll add a story in jira.

File: applications/asset_tracker/src/main.c [hunk 1]
```diff
@@ -239,6 +251,58 @@ static void send_button_data_work_fn(struct k_work *work)
...
+	if (err) {
+		snprintf(modem_at_cmd_buff, sizeof(modem_at_cmd_buff),
+				"AT CMD error: %d, state %d", err, state);
+	} else if (strlen(modem_at_cmd_buff) == 0) {
+		snprintf(modem_at_cmd_buff, sizeof(modem_at_cmd_buff),
+				"OK\r\n");
+	}
+
+	modem_data.data.buf = modem_at_cmd_buff;
+	modem_data.data.len = strlen(modem_at_cmd_buff);
```
Comment (jtguggedal) [inline applications/asset_tracker/src/main.c]: I think we need to cap the length at ~2k, as the modem can't handle bigger TLS packets. At least I think it applies to data being sent as well.
@rlubos Do you know?

File: applications/asset_tracker/src/main.c [hunk 2]
```diff
@@ -808,9 +907,12 @@ static void long_press_handler(struct k_work *work)
 /**@brief Initializes and submits delayed work. */
 static void work_init(void)
 {
+	k_sem_init(&modem_at_cmd_sem, 1, 1);
```
Comment (jtguggedal) [inline applications/asset_tracker/src/main.c]: Don't think this is needed when `K_SEM_DEFINE(...)` is used

File: applications/asset_tracker/src/main.c (hunk 1 above)
Comment (jayteemo) [reply applications/asset_tracker/src/main.c]: I assume the size limit goes both ways.  Either we have a local err  message as a result of the send failure, a local error message saying the at cmd response is too big, or we send a truncated at cmd response.  If the latter is preferred, what should the max length be, after the overhead?

File: applications/asset_tracker/src/cloud_codec/cloud_codec.c [hunk 3]
```diff
@@ -390,11 +414,10 @@ static int cloud_decode_modem_params(cJSON *const data_obj,
 	}
 
 	blob = json_object_decode(data_obj, MODEM_PARAM_BLOB_KEY_STR);
-	params->blob = cJSON_IsString(blob) ? blob->valuestring : NULL;
+	params->blob = cJSON_GetStringValue(blob);
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.c]: yes

File: applications/asset_tracker/src/cloud_codec/cloud_codec.h [hunk 4]
```diff
@@ -177,6 +182,7 @@ struct cloud_command {
 	union {
 		struct cloud_command_state_value sv;
 		struct cloud_command_modem_params mp;
+		char *data_string;
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply applications/asset_tracker/src/cloud_codec/cloud_codec.h]: Always null terminated as long as we are using the cJSON lib

File: applications/asset_tracker/src/main.c [hunk 5]
```diff
@@ -116,10 +117,21 @@ static struct k_work connect_work;
...
 static struct k_work device_status_work;
+
+#if defined(CONFIG_AT_CMD)
+#define MODEM_AT_CMD_BUFFER_LEN (CONFIG_AT_CMD_RESPONSE_MAX_LEN+1)
+#else
+#define MODEM_AT_CMD_NOT_ENABLED_STR "Error: AT Command driver is not enabled"
+#define MODEM_AT_CMD_BUFFER_LEN (sizeof(MODEM_AT_CMD_NOT_ENABLED_STR))
+#endif
+static char modem_at_cmd_buff[MODEM_AT_CMD_BUFFER_LEN];
+K_SEM_DEFINE(modem_at_cmd_sem, 1, 1);
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply applications/asset_tracker/src/main.c]: which?

[the remark this answers is not in the document]
Comment (jtguggedal) [reply applications/asset_tracker/src/main.c]: `static K_SEM_DEFINE()`, it's not static by default

File: applications/asset_tracker/src/main.c (hunk 1 above)
Comment (jtguggedal) [reply applications/asset_tracker/src/main.c]: I'd vote for sending an error message indicating response is too big, ie > 2kB. Can't think of much else than the CA certificate readout being bigger than the limit, though there are probably other commands that can produce giant responses.

File: applications/asset_tracker/src/main.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jayteemo) [reply applications/asset_tracker/src/main.c]: ah gotcha, I thought it was.  Will change.

**Pull request merged**